File: vnproglib-1.1/unity/examples/getting_started/Assets/Math/Attitude.cs
using System;

namespace VectorNav.Math
{

/// <summary>
/// Representation of an orientation/attitude.
/// </summary>
public struct AttitudeF
{
	#region Types

	private enum AttitudeType
	{
		YprRads,
		YprDegs,
		Quat,
		Dcm
	}

	#endregion

	#region Properties

	/// <summary>
	/// Returns an <c>AttitudeF</c> representing no rotation.
	/// </summary>
	public static AttitudeF NoRotation
	{
		get { return new AttitudeF(AttitudeType.Quat, quatf.NoRotation); }
	}

	/// <summary>
	/// Returns the orientation as represented in yaw, pitch, roll in degrees.
	/// </summary>
	public vec3f YprInDegs
	{
		get
		{
			switch (_underlyingType)
			{
				case AttitudeType.YprDegs:
					return (vec3f) _attitudeData;
				case AttitudeType.YprRads:
					return Conv.Rad2Deg((vec3f) _attitudeData);
				case AttitudeType.Quat:
					return Conv.Quat2YprInDegs((vec4f) _attitudeData);
				case AttitudeType.Dcm:
					return Conv.Dcm2YprInDegs((mat3f) _attitudeData);
				default:
					// Don't expect to ever get here.
					throw new NotImplementedException();
			}
		}
	}

	/// <summary>
	/// Returns the orientation as represented in yaw, pitch, roll in radians.
	/// </summary>
	public vec3f YprInRads
	{
		get
		{
			switch (_underlyingType)
			{
				case AttitudeType.YprRads:
					return (vec3f) _attitudeData;
				case AttitudeType.YprDegs:
					return Conv.Deg2Rad((vec3f) _attitudeData);
				case AttitudeType.Quat:
					return Conv.Quat2YprInRads((vec4f) _attitudeData);
				case AttitudeType.Dcm:
					return Conv.Dcm2YprInRads((mat3f) _attitudeData);
				default:
					// Don't expect to ever get here.
					throw new NotImplementedException();
			}
		}
	}

	/// <summary>
	/// Returns the orientation as represented in quaternion.
	/// </summary>
	public quatf Quat
	{
		get
		{
			switch (_underlyingType)
			{
				case AttitudeType.Quat:
					return (quatf) _attitudeData;
				case AttitudeType.YprDegs:
					return Conv.YprInDegs2Quat((vec3f) _attitudeData);
				case AttitudeType.YprRads:
					return Conv.YprInRads2Quat((vec3f) _attitudeData);
				case AttitudeType.Dcm:
					return Conv.Dcm2Quat((mat3f) _attitudeData);
				default:
					// Don't expect to ever get here.
					throw new NotImplementedException();
			}
		}
	}

	/// <summary>
	/// Returns the orientation as represented by a direction cosine matrix.
	/// </summary>
	public mat3f Dcm
	{
		get
		{
			switch (_underlyingType)
			{
				case AttitudeType.Dcm:
					return (mat3f) _attitudeData;
				case AttitudeType.YprDegs:
					return Conv.YprInDegs2Dcm((vec3f) _attitudeData);
				case AttitudeType.YprRads:
					return Conv.YprInRads2Dcm((vec3f) _attitudeData);
				case AttitudeType.Quat:
					return Conv.Quat2Dcm((quatf) _attitudeData);
				default:
					// Don't expect to ever get here.
					throw new NotImplementedException();
			}
		}
	}

	#endregion

	#region Constructors

	private AttitudeF(AttitudeType type, object attitude)
	{
		_underlyingType = type;
		_attitudeData = attitude;
	}

	#endregion

	#region Methods

	/// <summary>
	/// Creates a new <c>AttitudeF</c> from a quaternion.
	/// </summary>
	/// <param name="quat">
	/// The orientation expressed as a quaternion.
	/// </param>
	/// <returns>
	/// The new <c>AttitudeF</c>.
	/// </returns>
	public static AttitudeF FromQuat(quatf quat)
	{
		return new AttitudeF(AttitudeType.Quat, quat);
	}

	/// <summary>
	/// Creates a new <c>AttitudeF</c> from yaw, pitch, roll in degrees.
	/// </summary>
	/// <param name="ypr">
	/// The yaw, pitch, roll in degrees.
	/// </param>
	/// <returns>
	/// The new <c>AttitudeF</c>.
	/// </returns>
	public static AttitudeF FromYprInDegs(vec3f ypr)
	{
		return new AttitudeF(AttitudeType.YprDegs, ypr);
	}

	/// <summary>
	/// Creates a new <c>AttitudeF</c> from yaw, pitch, roll in radians.
	/// </summary>
	/// <param name="ypr">
	/// The yaw, pitch, roll in radians.
	/// </param>
	/// <returns>
	/// The new <c>AttitudeF</c>.
	/// </returns>
	public static AttitudeF FromYprInRads(vec3f ypr)
	{
		return new AttitudeF(AttitudeType.YprRads, ypr);
	}

	/// <summary>
	/// Creates a new <c>AttitudeF</c> from a direction cosine matrix.
	/// </summary>
	/// <param name="dcm">
	/// The direction cosine matrix.
	/// </param>
	/// <returns>
	/// The new <c>AttitudeF</c>.
	/// </returns>
	public static AttitudeF FromDcm(mat3f dcm)
	{
		return new AttitudeF(AttitudeType.Dcm, dcm);
	}

	#endregion

	private readonly AttitudeType _underlyingType;
	private readonly object _attitudeData;
}

}

File: vectornav/vnproglib-1.1.5.0/cpp/src/attitude.cpp
#include "vn/attitude.h"
#include "vn/exceptions.h"
#include "vn/conversions.h"

using namespace std;

namespace vn {
namespace math {

AttitudeF AttitudeF::noRotation()
{
	vec4f q(0, 0, 0, 1);

	return AttitudeF(ATT_Quat, &q);
}

AttitudeF AttitudeF::fromQuat(vec4f quat)
{
	return AttitudeF(ATT_Quat, &quat);
}

AttitudeF AttitudeF::fromYprInDegs(vec3f yprInDegs)
{
	return AttitudeF(ATT_YprDegs, &yprInDegs);
}

AttitudeF AttitudeF::fromYprInRads(vec3f yprInRads)
{
	return AttitudeF(ATT_YprRads, &yprInRads);
}

AttitudeF AttitudeF::fromDcm(mat3f dcm)
{
	return AttitudeF(ATT_Dcm, &dcm);
}

AttitudeF::AttitudeF(AttitudeType type, void* attitude) :
	_underlyingType(type)
{
	size_t numOfBytesToCopy = sizeof(vec3f);

	if (_underlyingType == ATT_Quat)
		numOfBytesToCopy = sizeof(vec4f);
	else if (_underlyingType == ATT_Dcm)
		numOfBytesToCopy = sizeof(mat3f);

	copy(static_cast<uint8_t*>(attitude), static_cast<uint8_t*>(attitude) + numOfBytesToCopy, _data);
}

vec3f AttitudeF::yprInDegs()
{
	switch (_underlyingType)
	{
	case ATT_YprDegs:
		return *static_cast<vec3f*>(static_cast<void*>(_data));
	case ATT_YprRads:
		return rad2deg(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_Quat:
		return quat2YprInDegs(*static_cast<vec4f*>(static_cast<void*>(_data)));
	case ATT_Dcm:
		return dcm2YprInDegs(*static_cast<mat3f*>(static_cast<void*>(_data)));
	default:
		throw not_implemented();
	}
}

vec3f AttitudeF::yprInRads()
{
	switch (_underlyingType)
	{
	case ATT_YprRads:
		return *static_cast<vec3f*>(static_cast<void*>(_data));
	case ATT_YprDegs:
		return deg2rad(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_Quat:
		return quat2YprInRads(*static_cast<vec4f*>(static_cast<void*>(_data)));
	case ATT_Dcm:
		return dcm2YprInRads(*static_cast<mat3f*>(static_cast<void*>(_data)));
	default:
		throw not_implemented();
	}
}

vec4f AttitudeF::quat()
{
	switch (_underlyingType)
	{
	case ATT_Quat:
		return *static_cast<vec4f*>(static_cast<void*>(_data));
	case ATT_YprDegs:
		return yprInDegs2Quat(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_YprRads:
		return yprInRads2Quat(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_Dcm:
		return dcm2quat(*static_cast<mat3f*>(static_cast<void*>(_data)));
	default:
		throw not_implemented();
	}
}

mat3f AttitudeF::dcm()
{
	switch (_underlyingType)
	{
	case ATT_Dcm:
		return *static_cast<mat3f*>(static_cast<void*>(_data));
	case ATT_YprDegs:
		return yprInDegs2Dcm(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_YprRads:
		return yprInRads2Dcm(*static_cast<vec3f*>(static_cast<void*>(_data)));
	case ATT_Quat:
		return quat2dcm(*static_cast<vec4f*>(static_cast<void*>(_data)));
	default:
		throw not_implemented();
	}
}

}
}
